[cae760] Frame has non-empty accessible name (#2034)
act-rules/act-rules.github.io: _rules/iframe-non-empty-accessible-name-cae760.md

  authors:
    - Jey Nandakumar
    - Wilco Fiers
    - Tom Brunet
  funding:
    - WAI-Tools
---

## Applicability

This rule applies to `iframe` elements that are [included in the accessibility tree][] and for which all of the following are true:

> - does not have a negative `tabindex` [attribute value][]; and
> - does not have a `role` [attribute value][] of `presentation`; and
> - does not have a `role` [attribute value][] of `none`.

Reviewer: I would suggest this:

```suggestion
- the `iframe` does not have a negative `tabindex` [attribute value][]; and
- the `iframe` does not have an [explicit semantic role][] of `presentation` or `none`.
```

Restated to follow the conditional list writing style, and then using explicit semantic role, so we don't require an exact attribute value match, to account for whitespace and fallback roles.

> **Note:** `frame` element is deprecated, this rule does not consider `frame` or `frameset` elements.

Reviewer: Would you mind moving this note to the background? That feels like a better place for it.

Author: Done

## Accessibility Support

- Browser and assistive technology support for `iframe` elements is currently **inconsistent**. Some examples of inconsistencies include (but are not limited to):
  - There is a known combination of a popular browser and assistive technology that ignores `aria-label` and only announces `title` attribute as an [accessible name][]
  - Some assistive technologies ignore empty `iframe` elements, regardless of if they are focusable or if they have an accessible name.
  - Some browsers instantly redirect focus from `iframe` elements  to the first focusable element inside that iframe. This redirect makes it appear as though the `iframe` never receives focus. This occurs even if the `iframe` has a non-negative `tabindex` [attribute value][].
  - Not all browsers redirect focus on `iframe` elements. This ensures that the contents of `iframe` elements can be scrolled and accessed by using the keyboard. This must not be circumvented by using a negative tabindex, as this will make the `iframe` completely inaccessible for keyboard navigation.

> Due to these inconsistencies, the applicability of this rule is scoped for `iframe` elements for which there is no indication of an attempt to remove them from accessibility consideration.  Whether `iframe` elements that are inapplicable to this rule still require an accessible name varies between browsers.

Reviewer 2: **Suggestion:** I feel this is more of a Background note than an Accessibility Support one 🤔 
Maybe phrasing it as:
```suggestion
Due to inconsistencies in handling focus on `iframe`, this rule ignores `iframe` elements for which there is an attempt to hide them from assistive technologies.  Whether `iframe` elements that are inapplicable to this rule still require an accessible name varies between browsers.
```
and move it to the Background section.
(also trying to simplify the phrasing and avoid negations)

Author: @WilcoFiers I like the rewording, but wanted to get your thoughts on Accessibility Support section vs Background before accepting this.

Reviewer: LGTM 👍

## Background

> **Note:** `frame` element is deprecated, this rule does not consider `frame` or `frameset` elements.

Reviewer 2: ```suggestion
The `frame` element is deprecated, this rule does not consider `frame` or `frameset` elements.
```
No need for the "note" marker in the Background section.

## Accessibility Support

- Browser and assistive technology support for `iframe` elements is currently **inconsistent**. Some examples of inconsistencies include (but are not limited to):
  - There is a known combination of a popular browser and assistive technology that ignores `aria-label` and only announces `title` attribute as an [accessible name][]
  - Some assistive technologies ignore empty `iframe` elements, regardless of if they are focusable or if they have an accessible name.

>   - Some browsers do not include `iframe` elements in the [sequential focus navigation][] if content of the `iframe` is in the [sequential focus navigation][], even if the `iframe` has a non-negative `tabindex` [attribute value][].

Reviewer: ```suggestion
  - Some browsers instantly redirect focus from `iframe` elements  to the first focusable element inside that iframe. This redirect makes it appear as though the `iframe` never receives focus. This occurs even if the `iframe` has a non-negative `tabindex` [attribute value][].
```

>   - Some browsers include `iframe` elements in the [sequential focus navigation][] unless the `iframe` is explicitly removed from [sequential focus navigation][] (by having the `tabindex` attribute set to a negative value). This ensures that the contents of `iframe` elements can be scrolled and accessed by using the keyboard.

Reviewer: I don't think we should mention this

```suggestion
  - Not all browsers redirect focus on `iframe` elements. This ensures that the contents of `iframe` elements can be scrolled and accessed by using the keyboard. This must not be circumvented by using a negative tabindex, as this will make the `iframe` completely inaccessible for keyboard navigation.
```

> Due to these inconsistencies, the applicability of this rule is scoped for `iframe` elements for which there is no indication of an attempt to remove them from accessibility consideration.

Reviewer: ```suggestion
Due to these inconsistencies, the applicability of this rule is scoped for `iframe` elements for which there is no indication of an attempt to remove them from accessibility consideration.  Whether `iframe` elements that are inapplicable to this rule still require an accessible name varies between browsers.
```

This `iframe` element has a negative `tabindex` [attribute value][].

```html
<iframe tabindex="-1" src="/test-assets/SC4-1-2-frame-doc.html" style="height: 250px"> </iframe>
```

#### Inapplicable Example 4

> This `iframe` element has a `role` [attribute value][] of `none`.

Reviewer 2: ```suggestion
This `iframe` element has an [explicit semantic role][] of `none`.
```
matching the newest Applicability.

> <iframe tabindex="-1" src="/test-assets/SC4-1-2-frame-doc.html"> </iframe>

Reviewer: Hmm, this one needs a height so that it isn't scrollable. 

https://www.w3.org/WAI/content-assets/wcag-act-rules/testcases/cae760/803b52f32a5d2193683f152aa5e00174d5f8416a.html

Author: Added

```

#### Inapplicable Example 4

This `iframe` element has a `role` [attribute value][] of `none`.

```html

> <iframe title=" " src="/test-assets/SC4-1-2-frame-doc.html" role="none"> </iframe>

Reviewer: This feels like a slightly better test of this. This could be passed because the title attribute is used.

```suggestion
<iframe src="/test-assets/SC4-1-2-frame-doc.html" role="none"> </iframe>
```
